client/views/view_units.cpp

```diff
@@ -434,7 +490,200 @@ void units_view::upgrade_units()
+          gold_cost += punit->upkeep[O_GOLD];
+          food_cost += punit->upkeep[O_FOOD];
+          shield_cost += punit->upkeep[O_SHIELD];
+        }
+      }
+    }
+    unit_list_iterate_end;
+
+    city_list_iterate(client_player()->cities, pcity)
+    {
+      if (pcity->production.value.utype == unittype
+          && pcity->production.kind == VUT_UTYPE) {
+        in_progress++;
+      }
+    }
+    city_list_iterate_end;
+
+    // Skip unused unit types
+    if (count == 0 && in_progress == 0) {
+      continue;
+    }
+
+    entries[*num_entries_used].type = unittype;
+    entries[*num_entries_used].count = count;
+    entries[*num_entries_used].in_prod = in_progress;
+    entries[*num_entries_used].upg = upgradable;
+    entries[*num_entries_used].gold_cost = gold_cost;
+    entries[*num_entries_used].food_cost = food_cost;
+    entries[*num_entries_used].shield_cost = shield_cost;
+    (*num_entries_used)++;
+  }
+  unit_type_iterate_end;
+
+  std::sort(entries, entries + *num_entries_used,
+            [](const auto &lhs, const auto &rhs) {
+              return QString::localeAwareCompare(
+                         utype_name_translation(lhs.type),
+                         utype_name_translation(rhs.type))
+                     < 0;
+            });
+
+  return {};
+}
+
+/**
+ * Returns an array of units subject to unitwaittime.
+ */
+std::vector<unit_waiting_entry>
+get_units_waiting_data(struct unit_waiting_entry *entries,
+                       int *num_entries_used)
+{
+  *num_entries_used = 0;
+
+  if (nullptr == client.conn.playing) {
+    return {};
+  }
+
+  unit_type_iterate(unittype)
+  {
+    int count = 0; // Count of active unit type
+    int pcity_near_dist = 0;
+    QString city_name;
+    time_t dt = 0;
+    char buf[64];
+
+    unit_list_iterate(client.conn.playing->units, punit)
+    {
+      struct city *pcity_near = get_nearest_city(punit, &pcity_near_dist);
+
+      if (unit_type_get(punit) == unittype && !can_unit_move_now(punit)
+          && punit->ssa_controller == SSA_NONE) {
+        count++;
+        city_name = get_nearest_city_text(pcity_near, pcity_near_dist);
+        dt = time(nullptr) - punit->action_timestamp;
+
+        if (dt != 0 && !can_unit_move_now(punit)) {
+          format_time_duration(-dt, buf, sizeof(buf));
+        }
+      }
+    }
+    unit_list_iterate_end;
+
+    // Skip unused unit types
+    if (count == 0) {
+      continue;
+    }
+
+    entries[*num_entries_used].type = unittype;
+    entries[*num_entries_used].city_name = city_name;
+    entries[*num_entries_used].timer = buf;
+
+    (*num_entries_used)++;
+  }
+  unit_type_iterate_end;
+
+  std::sort(entries, entries + *num_entries_used,
+            [](const auto &lhs, const auto &rhs) {
+              return QString::localeAwareCompare(
+                         utype_name_translation(lhs.type),
+                         utype_name_translation(rhs.type))
+                     < 0;
+            });
+
+  return {};
```

Comment:
Guess you shouldn't use `std::vector` if you're anyway always returning an empty one